napi-inl.h

```diff
@@ -336,14 +347,36 @@ inline double Number::DoubleValue() const {
 // String class
 ////////////////////////////////////////////////////////////////////////////////
 
-inline String String::New(napi_env env, const char* val, int length) {
+inline String String::New(napi_env env, const std::string& val) {
+  return String::New(env, val.c_str(), val.size());
+}
+
+inline String String::New(napi_env env, const std::u16string& val) {
+  return String::New(env, val.c_str(), val.size());
+}
+
+inline String String::New(napi_env env, const char* val) {
```

Comment:
The `napi_value` created here is not temporary. It gets wrapped in the `String` object that is returned from this function.

Also, this code actually is part of a commit that is from my PR https://github.com/nodejs/node-api/pull/18, so it might be better to review that over there, and look at the other commits building on it separately.